src/backend/crossterm.rs

```diff
@@ -24,36 +23,310 @@ use crate::{
+    /// let mut backend = CrosstermBackend::on_stderr()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    ///
+    /// [raw mode]: crate::backend#raw-mode
+    /// [alternate screen]: crate::backend#alternate-screen
+    pub fn on_stderr() -> Result<CrosstermBackend<Stderr>, io::Error> {
+        CrosstermBackend::new(io::stderr())
+            .with_raw_mode()?
+            .with_alternate_screen()
+    }
+}
+
+impl CrosstermBackend<Stdout> {
+    /// Creates a new `CrosstermBackend` using [`std::io::stdout()`] as the buffer.
+    ///
+    /// This method also enables [raw mode] and enters the [alternate screen].
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// # use ratatui::backend::{Backend, CrosstermBackend};
+    /// let mut backend = CrosstermBackend::on_stdout()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    ///
+    /// [raw mode]: crate::backend#raw-mode
+    /// [alternate screen]: crate::backend#alternate-screen
+    pub fn on_stdout() -> Result<CrosstermBackend<Stdout>, io::Error> {
+        CrosstermBackend::new(io::stdout())
+            .with_raw_mode()?
+            .with_alternate_screen()
+    }
+}
+
+impl Default for CrosstermBackend<Stderr> {
+    // impl Default for CrosstermBackend {
+    /// Creates a new `CrosstermBackend` using [`std::io::stderr()`] as the buffer.
+    ///
+    /// It will usually be more convenient to use [`CrosstermBackend::on_stderr()`], which enables
+    /// raw mode and enters the alternate screen.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// # use std::io::Stderr;
+    /// # use ratatui::backend::CrosstermBackend;
+    /// let mut backend = CrosstermBackend::<Stderr>::default()
+    ///     .with_raw_mode()?
+    ///     .with_alternate_screen()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    fn default() -> Self {
+        Self::new(io::stderr())
+    }
+}
+
+impl Default for CrosstermBackend<Stdout> {
+    // impl Default for CrosstermBackend {
+    /// Creates a new `CrosstermBackend` using [`std::io::stdout()`] as the buffer.
+    ///
+    /// It will usually be more convenient to use [`CrosstermBackend::on_stdout()`], which enables
+    /// raw mode and enters the alternate screen.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// # use std::io::Stdout;
+    /// # use ratatui::backend::{Backend, CrosstermBackend};
+    /// let mut backend = CrosstermBackend::<Stdout>::default()
+    ///     .with_raw_mode()?
+    ///     .with_alternate_screen()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    fn default() -> Self {
+        Self::new(io::stdout())
+    }
+}
+
+impl<W> Drop for CrosstermBackend<W>
+where
+    W: Write,
+{
+    /// Disables raw mode, and mouse capture, leaves the alternate screen and shows the cursor.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if it fails to disable raw mode, leave the alternate screen, disable
+    /// mouse capture, or show the cursor.
```

Comment:
Can we not panic here instead? Maybe just log a warning? I can anticipate all kinds of weird crashes in CI or edge cases like that where I think it'd be weird if it just panicked on drop.

Reply:
I started with a perspective that you were right - it's odd to panic during drop, but I think I've convinced myself that panicking is better.

The times when this will fail are likely the same as when enabling / entering fail, so it's likely that the app should have failed before the object is created and so this will never be hit in practice. But in a total edge case when this happens (likely on a CI where the app has intentionally ignored a previous error), we likely want the panic to happen because it lets the app's panic hook actually do something with it (like log to a file). Writing anything to the screen is a problem if the app is in raw mode (newlines don't get printed), or it's on the alternate screen (the output disappears altogether)

I had a brief play to look at how a backtrace for this would look like, and whether it would have any useful info that could help diagnose the issue.

```rust
    #[test]
    fn drop_simulated_failure() {
        let mut backend = CrosstermBackend::on_stdout().unwrap();
        let _ = backend.clear();
        eprintln!("foo");
    }
```
```rust
    fn drop(&mut self) {
        // snip actual impl
        Err::<(), &str>("").expect("simulated failure in drop");
    }
```
`RUST_BACKTRACE = 1 cargo test`
```
---- backend::crossterm::tests::drop_simulated_failure stdout ----
foo
thread 'backend::crossterm::tests::drop_simulated_failure' panicked at 'simulated failure in drop: ""', src/backend/crossterm.rs:330:29
stack backtrace:
   0: rust_begin_unwind
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:67:14
   2: core::result::unwrap_failed
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/result.rs:1651:5
   3: core::result::Result<T,E>::expect
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/result.rs:1033:23
   4: <ratatui::backend::crossterm::CrosstermBackend<W> as core::ops::drop::Drop>::drop
             at ./src/backend/crossterm.rs:330:9
   5: core::ptr::drop_in_place<ratatui::backend::crossterm::CrosstermBackend<std::io::stdio::Stdout>>
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ptr/mod.rs:497:1
   6: ratatui::backend::crossterm::tests::drop_simulated_failure
             at ./src/backend/crossterm.rs:562:5
   7: ratatui::backend::crossterm::tests::drop_simulated_failure::{{closure}}
             at ./src/backend/crossterm.rs:558:33
   8: core::ops::function::FnOnce::call_once
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ops/function.rs:250:5
   9: core::ops::function::FnOnce::call_once
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ops/function.rs:250:5
```
The info this adds over logging is where the drop occurred and the call pattern that got us there. I think that's probably a reasonable thing to happen.

Reply:
> Writing anything to the screen is a problem if the app is in raw mode (newlines don't get printed), or it's on the alternate screen (the output disappears altogether)

Yeah, which is why I was thinking of using log::warn and then it's on the users to setup logging. 

Btw, is this how the backtrace looks without a panic hook?

Reply:
`Drop` approach for cleaning up the terminal is definitely a good idea but yeah, I'm not sure about panicking here as well.

Reply:
I left the panic in because it allows the app writer to take action as a result of the app failing. They can log the problem, notify, restart, whatever. Panicking allows the user to log, but logging doesn't allow the app to panic, so it's the worse of the two choices objectively.

Reply:
We should never panic with the expectation that it can be caught and acted on. This is what Results are for. The pattern I have seen elsewhere is that there is an explicit close method on the object, which can be called if the user is interested in the result. If not, Drop will call the close too, and simply ignore the failure. Logging would probably be fine too. But to panic on this case is IMO very unidiomatic Rust, especially when done just for convenience.

https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html

First comment here:
https://www.reddit.com/r/rust/comments/9q3jqn/how_is_rust_safe_when_panics_can_happen_out_of/

Reply:
Thank you for the feedback. In addition to the links provided, I some time reading a bunch of rust lang users forum posts regarding panic during drop, and I'm convinced that those who mentioned that we should not panic are correct. Particularly https://users.rust-lang.org/t/return-error-condition-from-drop-impl/4933 was the best argument for this.
So our options become:
- ignore errors during drop - termion does this right now
- eprintln / log / trace any errors
- write a destroy method that returns `Result` and don't do anything in drop if that's been called
- write methods on terminal that handle reverting things nicely. E.g. a slight modification from https://users.rust-lang.org/t/return-error-condition-from-drop-impl/4933/4?u=joshka:
```rust
impl Terminal {
    pub fn draw(&mut self, ...) { ... }
    fn finalize(self) -> Result { ... }

    pub fn with<F: FnOnce(&mut Self) -> Result>(func: F) -> Result {
        let mut terminal = Terminal { ... };
        try!(func(&mut terminal))
        terminal.finalize()
    }
}
```
I'm going to play a bit more with these ideas